README.md
WiiDPaste 0.19-1
====================

**DIES IST EINE ALPHA-SOFTWARE - BENUTZUNG AUF EIGENE GEFAHR**

WiiDPaste, geforkt von ZeroBin, ist ein minimalistischer, Open-Source Pastebin-Service.
Der Server hat keine Ahnung von den "Pastes" die du verschickst. Daten werden im Browser mit 256-AES ver- und entschlüsselt

Dies ist eine deutsche Modifikation von ZeroBin, wie sie auf [WiiDatabase Paste](http://wiidatabase.de/paste) eingesetzt wird.

Projektseite von ZeroBin (englisch):
http://sebsauvage.net/wiki/doku.php?id=php:zerobin

## Installation
1. Kopiere alle Dateien in einen Ordner auf deinen Server
2. Rufe die URL im Browser auf
3. Fertig!

## Lizenzinformationen

Copyright (c) 2012 Sébastien SAUVAGE (sebsauvage.net)

This software is provided 'as-is', without any express or implied warranty.
In no event will the authors be held liable for any damages arising from 
the use of this software.

Permission is granted to anyone to use this software for any purpose, 
including commercial applications, and to alter it and redistribute it 
freely, subject to the following restrictions:

    1. The origin of this software must not be misrepresented; you must 
       not claim that you wrote the original software. If you use this 
       software in a product, an acknowledgment in the product documentation
       would be appreciated but is not required.

    2. Altered source versions must be plainly marked as such, and must 
       not be misrepresented as being the original software.

    3. This notice may not be removed or altered from any source distribution.


CHANGELOG.md
# ZeroBin version history #

  * **Alpha 0.8 (2012-04-11):**
    * Source code not published yet.
    * Interface completely redesigned. Icons added.
    * Now properly supports IE6/7 (ugly display, but it works. "Clone" button is disabled though.)
    * Added one level of depth for storage directories (This is better for higher load servers).
    * php version is now checked (min: 5.2.6)
    * Better checks on posted json data on server.
    * Added "1 year" expiration.
    * URLs are now converted to clickable links. This include http, https, ftp and magnet links.
    * Clickable links include ''rel="nofollow"'' to discourage SEO.
    * On my public service (http://sebsauvage.net/paste/)
      * All data will be deleted (you were warned - this is a test service)
      * Default paste expiration is now 1 month to prevent clogging-up my host.

  * **Alpha 0.9 (2012-04-11):**
    * Oh bummer... IE 8 is as shitty as IE6/7: Its does not seem to support ''white-space:pre-wrap'' correctly. I had to activate the special handling mode. I still have to test IE 9.

  * **Alpha 0.10 (2012-04-12):**
    * IE9 does not seem to correctly support ''pre-wrap'' either. Special handling mode activated for all version of IE<10. (Note: **ALL other browsers** correctly support this feature.) 

  * **Alpha 0.11 (2012-04-12):**
    * Automatically ignore parameters (such as &utm_source=...) added //after// the anchor by some stupid Web 2.0 services.
    * First public release.

  * **Alpha 0.12 (2012-04-18):**
    * **DISCUSSIONS !** Now you can enable discussions on your pastes. Of course, posted comments and nickname are also encrypted and the server cannot see them.
    * This feature implies a change in storage format. You will have to delete all previous pastes in your ZeroBin. 
    * Added [[php:vizhash_gd|Vizhash]] as avatars, so you can match posters IP addresses without revealing them. (Same image = same IP). Of course the IP address cannot be deduced from the Vizhash.
    * Remaining time before expiration is now displayed.
    * Explicit tags were added to CSS and jQuery selectors (eg. div#aaa instead of #aaa) to speed up browser. 
    * Better cleaning of the URL (to make sure the key is not broken by some stupid redirection service)

  * **Alpha 0.13 (2012-04-18):**
    * FIXED: ''imageantialias()'' call removed because it's not really usefull and can be a problem on most hosts (if GD is not compiled in php).
    * FIXED: $error not properly initialized in index.php

  * **Alpha 0.14 (2012-04-20):**
    * ADDED: GD presence is checked. 
    * CHANGED: Traffic limiter data files moved to data/ (→easier rights management)
    * ADDED: "Burn after reading" implemented. Opening the URL will display the paste and immediately destroy it on server.

  * **Alpha 0.15 (2012-04-20):**
    * FIXED: 2 minor corrections to avoid notices in php log.
    * FIXED: Sources converted to UTF-8.

  * **Alpha 0.15 (2012-04-20):**
    * FIXED: 2 minor corrections to avoid notices in php log.
    * FIXED: Sources converted to UTF-8.

  * **Alpha 0.16**:
    * FIXED minor php warnings.
    * FIXED: zerobin.js reformated and properly commented.
    * FIXED: Directory structure re-organized.
    * CHANGED: URL shortening button was removed. (It was bad for privacy.)

  * **Alpha 0.17 (2013-02-23)**:
    * ADDED: Deletion URL.
    * small refactoring.
    * improved regex checks.
    * larger server alt on installation.    

  * **Alpha 0.18 (2013-02-24)**:
    * ADDED: The resulting URL is automatically selected after pressing "Send". You just have to press CTRL+C.
    * ADDED: Automatic syntax highlighting for 53 languages using highlight.js
    * ADDED: "5 minutes" and "1 week" expirations.
    * ADDED: "Raw text" button.
    * jQuery upgraded to 1.9.1
    * sjcl upgraded to GitHub master 2013-02-23
    * base64.js upgraded to 1.7
    * FIXED: Dates in discussion are now proper local dates.
    * ADDED: Robot meta tags in HTML to prevent search engines indexing.
    * ADDED: Better json checking (including entropy).
    * ADDED: Added version to js/css assets URLs in order to prevent some abusive caches to serve an obsolete version of these files when ZeroBin is upgraded.
    * "Burn after reading" option has been moved out of Expiration combo to a separate checkbox. Reason is: You can prevent a read-once paste to be available ad vitam eternam on the net.

  * **Alpha 0.19 (2013-07-05)**:
    * Corrected XSS security flaw which affected IE<10. Other browsers were not affected.
    * Corrected spacing display in IE<10.
    
  * **Alpha 0.19-1 (2014-03-29)**:
    * Bugfix-Release des WiiDatabase Teams, um die Sicherheit zu erhöhen:
    * FIXED: Weiter eingedeutscht
    * ADDED: Social Media Links hinzugefügt
    * CHANGED: Haupt-CSS und Theme-CSS gesplittet (erlaubt bessere Anpassung, ein Light-Theme ist mit enthalten!)
    * FIXED: Typos gefixt
    * CHANGED: Wenn der Paste-Identifier falsch ist, wird nun eine "Invalide Daten" Meldung angezeigt, anstatt "falscher Token"
    * FIXED: Sicherheitsproblem gefixt: Wenn man HTML/Javascript gepastet hat und mit RAW-Text betrachtet hat, wurde das HTML/Javascript interpretiert, anstatt angezeigt, was ein enormes Sicherheitsproblem ist. Alle Pastes werden nun so angezeigt, wie sie gepastet wurden. Sie werden nicht mehr ausgeführt
    * ADDED: Stärkerer Server-Salt. Dies fixt das Problem, dass VizHash exploited werden kann
    * Mehrere Bugs gefixt (u.a. "Optionaler Nickname..." als Name bei Kommentaren. Hier wird nun automatisch dies durch "(Anonym)" ersetzt!
    * Mehrere Sicherheitsprobleme gefixt


lib/serversalt.php
<?php

// Generate a large random hexadecimal salt.
function generateRandomSalt()
{
    $randomSalt='';
    if (function_exists("mcrypt_create_iv"))
    {
        $randomSalt = bin2hex(mcrypt_create_iv(256, MCRYPT_DEV_URANDOM));
    }
    else // fallback to mt_rand()
    {
        for($i=0;$i<16;$i++) { $randomSalt.=base_convert(mt_rand(),10,16); }
    }
    return $randomSalt;
}

/* Return this ZeroBin server salt.
   This is a random string which is unique to each ZeroBin installation.
   It is automatically created if not present.

   Salt is used:
      - to generate unique VizHash in discussions (which are not reproductible across ZeroBin servers)
      - to generate unique deletion token (which are not re-usable across ZeroBin servers)
*/
function getServerSalt()
{
    $saltfile = 'data/salt.php';
    if (!is_file($saltfile))
        file_put_contents($saltfile,'<?php /* |'.generateRandomSalt().'| */ ?>');
    $items=explode('|',file_get_contents($saltfile));
    return $items[1];

}

?>


lib/vizhash_gd_zero.php
<?php
// VizHash_GD 0.0.4 beta ZeroBin 0.19-1
// Visual Hash implementation in php4+GD, stripped down and modified version for ZeroBin
// See: http://sebsauvage.net/wiki/doku.php?id=php:vizhash_gd
// This is free software under the zlib/libpng licence
// http://www.opensource.org/licenses/zlib-license.php
/* Example:
    $vz = new vizhash16x16();
    $data = $vz->generate('hello');
    header('Content-type: image/png');
    echo $data;
    exit;
*/
require_once "serversalt.php";
class vizhash16x16
{
    private $VALUES;
    private $VALUES_INDEX;
    private $width;
    private $height;
    private $salt;
    function __construct()
    {
        $this->width=16;
        $this->height=16;
        $this->salt = getServerSalt();
    }  
    
    // Generate a 16x16 png corresponding to $text.
    // Input: $text (string)
    // Output: PNG data. Or empty string if GD is not available.
    function generate($text)
    {
        if (!function_exists('gd_info')) return '';

        // We hash the input string.
        $hash=hash('sha1',$text.$this->salt).hash('md5',$text.$this->salt);
        $hash=$hash.strrev($hash);  # more data to make graphics

        // We convert the hash into an array of integers.
        $this->VALUES=array();
        for($i=0; $i<strlen($hash); $i=$i+2){ array_push($this->VALUES,hexdec(substr($hash,$i,2))); }
        $this->VALUES_INDEX=0; // to walk the array.

        // Then use these integers to drive the creation of an image.
        $image = imagecreatetruecolor($this->width,$this->height);

        $r0 = $this->getInt();$r=$r0;
        $g0 = $this->getInt();$g=$g0;
        $b0 = $this->getInt();$b=$b0;

        // First, create an image with a specific gradient background.
        $op='v'; if (($this->getInt()%2)==0) { $op='h'; };
        $image = $this->degrade($image,$op,array($r0,$g0,$b0),array(0,0,0));

        for($i=0; $i<7; $i=$i+1)
        {     
            $action=$this->getInt();
            $color = imagecolorallocate($image, $r,$g,$b);
            $r = ($r0 + $this->getInt()/25)%256;
            $g = ($g0 + $this->getInt()/25)%256;
            $b = ($b0 + $this->getInt()/25)%256;
            $r0=$r; $g0=$g; $b0=$b;
            $this->drawshape($image,$action,$color);   
        }

        $color = imagecolorallocate($image,$this->getInt(),$this->getInt(),$this->getInt());
        $this->drawshape($image,$this->getInt(),$color);
        ob_start();
        imagepng($image);
        $imagedata = ob_get_contents();
        ob_end_clean();
        imagedestroy($image);
        
        return $imagedata;
    } 

    private function getInt() // Returns a single integer from the $VALUES array (0...255)
    {
        $v= $this->VALUES[$this->VALUES_INDEX]; 
        $this->VALUES_INDEX++;
        $this->VALUES_INDEX %= count($this->VALUES); // Warp around the array
        return $v;
    }
    private function getX() // Returns a single integer from the array (roughly mapped to image width) 
    {
        return $this->width*$this->getInt()/256;
    }

    private function getY() // Returns a single integer from the array (roughly mapped to image height) 
    { 
        return $this->height*$this->getInt()/256;
    }  
    
    # Gradient function taken from:
    # http://www.supportduweb.com/scripts_tutoriaux-code-source-41-gd-faire-un-degrade-en-php-gd-fonction-degrade-imagerie.html
    private function degrade($img,$direction,$color1,$color2)
    {
            if($direction=='h') { $size = imagesx($img); $sizeinv = imagesy($img); }
            else { $size = imagesy($img); $sizeinv = imagesx($img);}
            $diffs = array(
                    (($color2[0]-$color1[0])/$size),
                    (($color2[1]-$color1[1])/$size),
                    (($color2[2]-$color1[2])/$size)
            );
            for($i=0;$i<$size;$i++)
            {
                    $r = $color1[0]+($diffs[0]*$i);
                    $g = $color1[1]+($diffs[1]*$i);
                    $b = $color1[2]+($diffs[2]*$i);
                    if($direction=='h') { imageline($img,$i,0,$i,$sizeinv,imagecolorallocate($img,$r,$g,$b)); }
                    else { imageline($img,0,$i,$sizeinv,$i,imagecolorallocate($img,$r,$g,$b)); }
            }
            return $img;
    }
    
    private function drawshape($image,$action,$color)
    {
        switch($action%7)
        {
            case 0:
                ImageFilledRectangle ($image,$this->getX(),$this->getY(),$this->getX(),$this->getY(),$color);  
                break;
            case 1:
            case 2:
                ImageFilledEllipse ($image, $this->getX(), $this->getY(), $this->getX(), $this->getY(), $color);  
                break;
            case 3:
                $points = array($this->getX(), $this->getY(), $this->getX(), $this->getY(), $this->getX(), $this->getY(),$this->getX(), $this->getY());
                ImageFilledPolygon ($image, $points, 4, $color);
                break;
            case 4:
            case 5:
            case 6:
                $start=$this->getInt()*360/256; $end=$start+$this->getInt()*180/256;
                ImageFilledArc ($image, $this->getX(), $this->getY(), $this->getX(), $this->getY(),$start,$end,$color,IMG_ARC_PIE);
                break;     
        }
    }    
}    

?>


index.php
<?php
/*
ZeroBin - a zero-knowledge paste bin
Please see project page: http://sebsauvage.net/wiki/doku.php?id=php:zerobin
*/
$VERSION='Alpha 0.19-1';
if (version_compare(PHP_VERSION, '5.2.6') < 0) die('ZeroBin requires php 5.2.6 or above to work. Sorry.');
require_once "lib/serversalt.php";
require_once "lib/vizhash_gd_zero.php";

// In case stupid admin has left magic_quotes enabled in php.ini:
if (get_magic_quotes_gpc())
{
    function stripslashes_deep($value) { $value = is_array($value) ? array_map('stripslashes_deep', $value) : stripslashes($value); return $value; }
    $_POST = array_map('stripslashes_deep', $_POST);
    $_GET = array_map('stripslashes_deep', $_GET);
    $_COOKIE = array_map('stripslashes_deep', $_COOKIE);
}

// trafic_limiter : Make sure the IP address makes at most 1 request every 10 seconds.
// Will return false if IP address made a call less than 10 seconds ago.
function trafic_limiter_canPass($ip)
{
    $tfilename='./data/trafic_limiter.php';
    if (!is_file($tfilename))
    {
        file_put_contents($tfilename,"<?php\n\$GLOBALS['trafic_limiter']=array();\n?>", LOCK_EX);
        chmod($tfilename,0705);
    }
    require $tfilename;
    $tl=$GLOBALS['trafic_limiter'];
    if (!empty($tl[$ip]) && ($tl[$ip]+10>=time()))
    {
        return false;
        // FIXME: purge file of expired IPs to keep it small
    }
    $tl[$ip]=time();
file_put_contents($tfilename, "<?php\n\$GLOBALS['trafic_limiter']=".var_export($tl,true).";\n?>", LOCK_EX);
    return true;
}

// Constant time string comparison.
// (Used to deter time attacks on hmac checking. See section 2.7 of https://defuse.ca/audits/zerobin.htm)
function slow_equals($a, $b)
{
    $diff = strlen($a) ^ strlen($b);
    for($i = 0; $i < strlen($a) && $i < strlen($b); $i++)
    {
        $diff |= ord($a[$i]) ^ ord($b[$i]);
    }
    return $diff === 0;
}

/* Convert paste id to storage path.
   The idea is to creates subdirectories in order to limit the number of files per directory.
   (A high number of files in a single directory can slow things down.)
   eg. "f468483c313401e8" will be stored in "data/f4/68/f468483c313401e8"
   High-trafic websites may want to deepen the directory structure (like Squid does).

   eg. input 'e3570978f9e4aa90' --> output 'data/e3/57/'
*/
function dataid2path($dataid)
{
    return 'data/'.substr($dataid,0,2).'/'.substr($dataid,2,2).'/';
}

/* Convert paste id to discussion storage path.
   eg. 'e3570978f9e4aa90' --> 'data/e3/57/e3570978f9e4aa90.discussion/'
*/
function dataid2discussionpath($dataid)
{
    return dataid2path($dataid).$dataid.'.discussion/';
}

// Checks if a json string is a proper SJCL encrypted message.
// False if format is incorrect.
function validSJCL($jsonstring)
{
    $accepted_keys=array('iv','v','iter','ks','ts','mode','adata','cipher','salt','ct');

    // Make sure content is valid json
    $decoded = json_decode($jsonstring);
    if ($decoded==null) return false;
    $decoded = (array)$decoded;

    // Make sure required fields are present
    foreach($accepted_keys as $k)
    {
        if (!array_key_exists($k,$decoded))  { return false; }
    }

    // Make sure some fields are base64 data
    if (base64_decode($decoded['iv'],$strict=true)==null) { return false; }
    if (base64_decode($decoded['salt'],$strict=true)==null) { return false; }
    if (base64_decode($decoded['cipher'],$strict=true)==null) { return false; }

    // Make sure no additionnal keys were added.
    if (count(array_intersect(array_keys($decoded),$accepted_keys))!=10) { return false; }

    // Reject data if entropy is too low
    $ct = base64_decode($decoded['ct'], $strict=true);
    if (strlen($ct) > strlen(gzdeflate($ct))) return false;

    // Make sure some fields have a reasonable size.
    if (strlen($decoded['iv'])>24) return false;
    if (strlen($decoded['salt'])>14) return false;
    return true;
}

// Delete a paste and its discussion.
// Input: $pasteid : the paste identifier.
function deletePaste($pasteid)
{
    // Delete the paste itself
    unlink(dataid2path($pasteid).$pasteid);

    // Delete discussion if it exists.
    $discdir = dataid2discussionpath($pasteid);
    if (is_dir($discdir))
    {
        // Delete all files in discussion directory
        $dhandle = opendir($discdir);
        while (false !== ($filename = readdir($dhandle)))
        {
            if (is_file($discdir.$filename))  unlink($discdir.$filename);
        }
        closedir($dhandle);

        // Delete the discussion directory.
        rmdir($discdir);
    }
}

if (!empty($_POST['data'])) // Create new paste/comment
{
    /* POST contains:
         data (mandatory) = json encoded SJCL encrypted text (containing keys: iv,salt,ct)

         All optional data will go to meta information:
         expire (optional) = expiration delay (never,5min,10min,1hour,1day,1week,1month,1year,burn) (default:never)
         opendiscusssion (optional) = is the discussion allowed on this paste ? (0/1) (default:0)
         syntaxcoloring (optional) = should this paste use syntax coloring when displaying.
         nickname (optional) = son encoded SJCL encrypted text nickname of author of comment (containing keys: iv,salt,ct)
         parentid (optional) = in discussion, which comment this comment replies to.
         pasteid (optional) = in discussion, which paste this comment belongs to.
    */

    header('Content-type: application/json');
    $error = false;

    // Create storage directory if it does not exist.
    if (!is_dir('data'))
    {
        mkdir('data',0705);
        file_put_contents('data/.htaccess',"Allow from none\nDeny from all\n", LOCK_EX);
    }

    // Make sure last paste from the IP address was more than 10 seconds ago.
    if (!trafic_limiter_canPass($_SERVER['REMOTE_ADDR']))
        { echo json_encode(array('status'=>1,'message'=>'Bitte warte 10 Sekunden nach jedem Paste!')); exit; }

    // Make sure content is not too big.
    $data = $_POST['data'];
    if (strlen($data)>2000000)
        { echo json_encode(array('status'=>1,'message'=>'Dieser Paste ist grö&szlig;er als 2 MB!')); exit; }

    // Make sure format is correct.
    if (!validSJCL($data))
        { echo json_encode(array('status'=>1,'message'=>'Invalide Daten.')); exit; }

    // Read additional meta-information.
    $meta=array();

    // Read expiration date
    if (!empty($_POST['expire']))
    {
        $expire=$_POST['expire'];
        if ($expire=='5min') $meta['expire_date']=time()+5*60;
        elseif ($expire=='10min') $meta['expire_date']=time()+10*60;
        elseif ($expire=='1hour') $meta['expire_date']=time()+60*60;
        elseif ($expire=='1day') $meta['expire_date']=time()+24*60*60;
        elseif ($expire=='1week') $meta['expire_date']=time()+7*24*60*60;
        elseif ($expire=='1month') $meta['expire_date']=time()+30*24*60*60; // Well this is not *exactly* one month, it's 30 days.
        elseif ($expire=='1year') $meta['expire_date']=time()+365*24*60*60;
    }

    // Destroy the paste when it is read.
    if (!empty($_POST['burnafterreading']))
    {
        $burnafterreading = $_POST['burnafterreading'];
        if ($burnafterreading!='0' && $burnafterreading!='1') { $error=true; }
        if ($burnafterreading!='0') { $meta['burnafterreading']=true; }
    }

    // Read open discussion flag
    if (!empty($_POST['opendiscussion']))
    {
        $opendiscussion = $_POST['opendiscussion'];
        if ($opendiscussion!='0' && $opendiscussion!='1') { $error=true; }
        if ($opendiscussion!='0') { $meta['opendiscussion']=true; }
    }

    // Should we use syntax coloring when displaying ?
    if (!empty($_POST['syntaxcoloring']))
    {
        $syntaxcoloring = $_POST['syntaxcoloring'];
        if ($syntaxcoloring!='0' && $syntaxcoloring!='1') { $error=true; }
        if ($syntaxcoloring!='0') { $meta['syntaxcoloring']=true; }
    }    

    // You can't have an open discussion on a "Burn after reading" paste:
    if (isset($meta['burnafterreading'])) unset($meta['opendiscussion']);

    // Optional nickname for comments
    if (!empty($_POST['nickname']))
    {
        $nick = $_POST['nickname'];
        if (!validSJCL($nick))
        {
            $error=true;
        }
        else
        {
            $meta['nickname']=$nick;

            // Generation of the anonymous avatar (Vizhash):
            // If a nickname is provided, we generate a Vizhash.
            // (We assume that if the user did not enter a nickname, he/she wants
            // to be anonymous and we will not generate the vizhash.)
            $vz = new vizhash16x16();
            $pngdata = $vz->generate($_SERVER['REMOTE_ADDR']);
            if ($pngdata!='') $meta['vizhash'] = 'data:image/png;base64,'.base64_encode($pngdata);
            // Once the avatar is generated, we do not keep the IP address, nor its hash.
        }
    }

    if ($error)
    {
        echo json_encode(array('status'=>1,'message'=>'Invalide Daten.'));
        exit;
    }

    // Add post date to meta.
    $meta['postdate']=time();

    // We just want a small hash to avoid collisions: Half-MD5 (64 bits) will do the trick.
    $dataid = substr(hash('md5',$data),0,16);

    $is_comment = (!empty($_POST['parentid']) && !empty($_POST['pasteid'])); // Is this post a comment ?
    $storage = array('data'=>$data);
    if (count($meta)>0) $storage['meta'] = $meta;  // Add meta-information only if necessary.

    if ($is_comment) // The user posts a comment.
    {
        $pasteid = $_POST['pasteid'];
        $parentid = $_POST['parentid'];
        if (!preg_match('/\A[a-f\d]{16}\z/',$pasteid)) { echo json_encode(array('status'=>1,'message'=>'Invalide Daten.')); exit; }
        if (!preg_match('/\A[a-f\d]{16}\z/',$parentid)) { echo json_encode(array('status'=>1,'message'=>'Invalide Daten.')); exit; }

        unset($storage['expire_date']); // Comment do not expire (it's the paste that expires)
        unset($storage['opendiscussion']);
        unset($storage['syntaxcoloring']);

        // Make sure paste exists.
        $storagedir = dataid2path($pasteid);
        if (!is_file($storagedir.$pasteid)) { echo json_encode(array('status'=>1,'message'=>'Invalide Daten.')); exit; }

        // Make sure the discussion is opened in this paste.
        $paste=json_decode(file_get_contents($storagedir.$pasteid));
        if (!$paste->meta->opendiscussion) { echo json_encode(array('status'=>1,'message'=>'Invalide Daten.')); exit; }

        $discdir = dataid2discussionpath($pasteid);
        $filename = $pasteid.'.'.$dataid.'.'.$parentid;
        if (!is_dir($discdir)) mkdir($discdir,$mode=0705,$recursive=true);
        if (is_file($discdir.$filename)) // Oups... improbable collision.
        {
            echo json_encode(array('status'=>1,'message'=>'Du hast kein Glück. Versuche es erneut.'));
            exit;
        }

        file_put_contents($discdir.$filename,json_encode($storage), LOCK_EX);
        echo json_encode(array('status'=>0,'id'=>$dataid)); // 0 = no error
        exit;
    }
    else // a standard paste.
    {
        $storagedir = dataid2path($dataid);
        if (!is_dir($storagedir)) mkdir($storagedir,$mode=0705,$recursive=true);
        if (is_file($storagedir.$dataid)) // Oups... improbable collision.
        {
            echo json_encode(array('status'=>1,'message'=>'Du hast kein Glück. Versuche es erneut...'));
            exit;
        }
        // New paste
        file_put_contents($storagedir.$dataid,json_encode($storage), LOCK_EX);

        // Generate the "delete" token.
        // The token is the hmac of the pasteid signed with the server salt.
        // The paste can be delete by calling http://myserver.com/zerobin/?pasteid=<pasteid>&deletetoken=<deletetoken>
        $deletetoken = hash_hmac('sha1', $dataid , getServerSalt());

        echo json_encode(array('status'=>0,'id'=>$dataid,'deletetoken'=>$deletetoken)); // 0 = no error
        exit;
    }

echo json_encode(array('status'=>1,'message'=>'Server error.'));
exit;
}

/* Process a paste deletion request.
   Returns an array ('',$ERRORMESSAGE,$STATUS)
*/
function processPasteDelete($pasteid,$deletetoken)
{
    if (preg_match('/\A[a-f\d]{16}\z/',$pasteid))  // Is this a valid paste identifier ?
    {
        $filename = dataid2path($pasteid).$pasteid;
        if (!is_file($filename)) // Check that paste exists.
        {
            return array('','Paste existiert nicht, ist ausgelaufen oder wurde gelöscht.','');
        }
    }
    else
    {
        return array('','Invalide Daten','');
    }

    if ($deletetoken != hash_hmac('sha1', $pasteid , getServerSalt())) // Make sure token is valid.
    {
        return array('','Falscher Lösch-Token. Paste wurde nicht gelöscht.','');
    }

    // Paste exists and deletion token is valid: Delete the paste.
    deletePaste($pasteid);
    return array('','','Paste wurde erfolgreich gelöscht!');
}

/* Process a paste fetch request.
   Returns an array ($CIPHERDATA,$ERRORMESSAGE,$STATUS)
*/
function processPasteFetch($pasteid)
{
    if (preg_match('/\A[a-f\d]{16}\z/',$pasteid))  // Is this a valid paste identifier ?
    {
        $filename = dataid2path($pasteid).$pasteid;
        if (!is_file($filename)) // Check that paste exists.
        {
            return array('','Paste does not exist, has expired or has been deleted.','');
        }
    }
    else
    {
        return array('','Invalide Daten','');
    } 

    // Get the paste itself.
    $paste=json_decode(file_get_contents($filename));

    // See if paste has expired.
    if (isset($paste->meta->expire_date) && $paste->meta->expire_date<time())
    {
        deletePaste($pasteid);  // Delete the paste
        return array('','Paste existiert nicht, ist ausgelaufen oder wurde gelöscht.','');
    }


    // We kindly provide the remaining time before expiration (in seconds)
    if (property_exists($paste->meta, 'expire_date')) $paste->meta->remaining_time = $paste->meta->expire_date - time();

    $messages = array($paste); // The paste itself is the first in the list of encrypted messages.
    // If it's a discussion, get all comments.
    if (property_exists($paste->meta, 'opendiscussion') && $paste->meta->opendiscussion)
    {
        $comments=array();
        $datadir = dataid2discussionpath($pasteid);
        if (!is_dir($datadir)) mkdir($datadir,$mode=0705,$recursive=true);
        $dhandle = opendir($datadir);
        while (false !== ($filename = readdir($dhandle)))
        {
            if (is_file($datadir.$filename))
            {
                $comment=json_decode(file_get_contents($datadir.$filename));
                // Filename is in the form pasteid.commentid.parentid:
                // - pasteid is the paste this reply belongs to.
                // - commentid is the comment identifier itself.
                // - parentid is the comment this comment replies to (It can be pasteid)
                $items=explode('.',$filename);
                $comment->meta->commentid=$items[1]; // Add some meta information not contained in file.
                $comment->meta->parentid=$items[2];
                $comments[$comment->meta->postdate]=$comment; // Store in table
            }
        }
        closedir($dhandle);
        ksort($comments); // Sort comments by date, oldest first.
        $messages = array_merge($messages, $comments);
    }
    $CIPHERDATA = json_encode($messages);

    // If the paste was meant to be read only once, delete it.
    if (property_exists($paste->meta, 'burnafterreading') && $paste->meta->burnafterreading) deletePaste($pasteid);

    return array($CIPHERDATA,'','');
}



$CIPHERDATA='';
$ERRORMESSAGE='';
$STATUS='';

if (!empty($_GET['deletetoken']) && !empty($_GET['pasteid'])) // Delete an existing paste
{
    list ($CIPHERDATA, $ERRORMESSAGE, $STATUS) = processPasteDelete($_GET['pasteid'],$_GET['deletetoken']);
}
else if (!empty($_SERVER['QUERY_STRING']))  // Return an existing paste.
{
    list ($CIPHERDATA, $ERRORMESSAGE, $STATUS) = processPasteFetch($_SERVER['QUERY_STRING']);    

}

require_once "lib/rain.tpl.class.php";
header('Content-Type: text/html; charset=utf-8');
$page = new RainTPL;
$page->assign('CIPHERDATA',htmlspecialchars($CIPHERDATA,ENT_NOQUOTES));  // We escape it here because ENT_NOQUOTES can't be used in RainTPL templates.
$page->assign('VERSION',$VERSION);
$page->assign('ERRORMESSAGE',$ERRORMESSAGE);
$page->assign('STATUS',$STATUS);
$page->draw('page');
?>
